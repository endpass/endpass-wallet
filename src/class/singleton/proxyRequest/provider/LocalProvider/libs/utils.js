import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
