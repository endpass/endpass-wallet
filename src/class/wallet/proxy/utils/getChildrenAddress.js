export default function(hdWallet, offset, limit) {
  return [...Array(limit)].map((_, i) =>
    hdWallet
      .deriveChild(offset + i)
      .getWallet()
      .getChecksumAddressString(),
  );
}
