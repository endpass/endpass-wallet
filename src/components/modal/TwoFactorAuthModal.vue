<template lang="html">
  <v-modal
    class="is-dark"
    @close="close"
  >
    <template slot="header">Two Factor Authentication</template>
    <div v-if="secret">
      <p class="lead">To protect your account, we strongly recommend enabling
      two factor authentication. Once enabled, you will be able to log in by
      entering a code generated by an authenticator app.</p>

      <p class="lead">Scan the QR code below into <strong>Google
      Authenticator</strong>. If you are unable to scan the code, you can
      also enter the secret manually.</p>

    </div>
    <v-form
      id="twoFactorAuth"
      @submit="confirm"
    >
      <div
        v-if="secret && email"
        class="field"
      >
        <div class="has-text-centered">
          <img :src="qrCodeSrc">
        </div>
        <p class="subtitle">Write down the recovery secret below:</p>
        <label class="label">Secret</label>
        <p class="subtitle twofa-secret">
          <strong>{{ secret }}</strong>
        </p>
      </div>
      <p class="subtitle">Enter the 6 digit code generated by your
      authenticator app.</p>
      <v-input
        v-model="code"
        type="number"
        size="6"
        label="Verification Code"
        name="verificationCode"
        validator="required|digits:6"
      />
    </v-form>
    <div
      slot="footer"
      class="buttons"
    >
      <v-button
        :loading="isLoading"
        form="twoFactorAuth"
        class-name="is-primary is-medium"
      >
        Verify
      </v-button>
    </div>
  </v-modal>
</template>

<script>
import VModal from '@/components/ui/VModal';
import VForm from '@/components/ui/form/VForm';
import VInput from '@/components/ui/form/VInput';
import VButton from '@/components/ui/form/VButton';

export default {
  // TODO: must be in PascalCase, breaks tests
  name: 'two-factor-auth-modal',
  props: {
    secret: String,
    email: String,
    isLoading: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      code: null,
    };
  },
  computed: {
    qrCodeSrc() {
      const otpAuthUri = `otpauth://totp/Endpass:${this.email}?secret=${
        this.secret
      }`;

      return `https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=${otpAuthUri}`;
    },
  },
  methods: {
    confirm() {
      this.$emit('confirm', this.code);
    },
    close() {
      this.$emit('close');
    },
  },
  components: {
    VModal,
    VForm,
    VInput,
    VButton,
  },
};
</script>
