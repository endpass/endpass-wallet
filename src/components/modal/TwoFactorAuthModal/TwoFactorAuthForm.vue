<template>
  <v-form
    id="twoFactorAuthForm"
    :is-form-valid="isFormValid"
    @submit.native.prevent="onFormSubmit"
  >
    <div
      v-if="isShowQRCode"
      class="two-factor-auth-form-section"
      data-test="two-factor-modal-qr-code"
    >
      <div class="has-text-centered">
        <img :src="qrCodeSrc">
      </div>
      <label class="subtitle">
        {{ $t('components.twoFactorAuthModal.writeDownCode') }}
      </label>
      <label class="label">
        {{ $t('components.twoFactorAuthModal.secret') }}
      </label>
      <p class="subtitle twofa-secret">
        <strong>{{ secret }}</strong>
      </p>
    </div>
    <div class="two-factor-auth-form-section">
      <label class="subtitle">
        {{ $t('components.twoFactorAuthModal.enterCode') }}
      </label>
      <v-input
        v-model="code"
        v-validate="'required|digits:6'"
        :error="errors.first('code')"
        type="number"
        size="6"
        :label="$t('components.twoFactorAuthModal.otpCode')"
        name="code"
        data-vv-name="code"
        data-test="input-two-auth-code"
        autofocus
        @input="handleInput"
      />
    </div>
    <v-button
      :loading="isLoading"
      :disabled="!isFormValid"
      type="submit"
      class-name="is-primary is-medium"
      data-test="submit-two-auth-modal"
    >
      {{ $t('global.verify') }}
    </v-button>
  </v-form>
</template>

<script>
import formMixin from '@/mixins/form';

export default {
  name: 'TwoFactorAuthForm',

  props: {
    secret: {
      type: String,
      default: '',
    },
    email: {
      type: String,
      default: '',
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    code: '',
  }),

  computed: {
    isShowQRCode() {
      return this.email && this.secret;
    },

    qrCodeSrc() {
      const otpAuthUri = `otpauth://totp/Endpass:${this.email}?issuer=Endpass&secret=${this.secret}`;

      return `https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=${encodeURIComponent(
        otpAuthUri,
      )}`;
    },
  },

  methods: {
    onFormSubmit() {
      if (!this.isFormValid) return;

      this.$emit('submit', this.code);
    },
  },

  mixins: [formMixin],
};
</script>

<style lang="scss">
.two-factor-auth-form-section {
  margin-bottom: 30px;
}
</style>
