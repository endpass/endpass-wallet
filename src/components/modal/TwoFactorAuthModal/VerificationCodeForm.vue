<template>
  <v-form
    id="verificationCode"
    :is-form-valid="isFormValid"
    @submit="onFormSubmit"
  >
    <div class="verification-code-form-section">
      <label class="subtitle">
        {{ $t('components.twoFactorAuthModal.enterVerificationCode') }}
      </label>
      <v-input
        v-model="verificationCode"
        v-validate="'required|digits:6'"
        :error="errors.first('verificationCode')"
        type="number"
        size="6"
        :label="$t('components.twoFactorAuthModal.verificationCode')"
        name="verificationCode"
        data-vv-name="verificationCode"
        data-test="input-two-auth-verification-code"
        autofocus
      />
    </div>
    <div class="verification-code-form-section v-text-center">
      <span>{{ $t('components.twoFactorAuthModal.didNotGetCode') }}</span>
      <a class="link" @click="onCodeRequest">
        {{ $t('components.twoFactorAuthModal.resendCode') }}
      </a>
    </div>
    <v-button
      :loading="isLoading"
      :disabled="!isFormValid"
      class-name="is-primary is-medium"
      data-test="submit-two-auth-modal"
    >
      {{ $t('global.continue') }}
    </v-button>
  </v-form>
</template>

<script>
import formMixin from '@/mixins/form';

export default {
  name: 'VerificationCodeForm',

  props: {
    email: {
      type: String,
      default: '',
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    verificationCode: '',
  }),

  methods: {
    onFormSubmit() {
      if (!this.isFormValid) return;

      this.$emit('submit', this.verificationCode);
    },

    onCodeRequest() {
      this.$emit('request-code');
    },
  },

  mixins: [formMixin],
};
</script>

<style lang="scss">
.verification-code-form-section {
  margin-bottom: 30px;
}
</style>
