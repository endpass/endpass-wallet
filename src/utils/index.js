import store from '../store';

export { default as http } from './http';

export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());

export const getInitializedValueFromStore = (module, field) => {
  return new Promise(resolve => {
    const value = module[field];

    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
};
