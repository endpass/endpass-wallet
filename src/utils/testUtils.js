import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
