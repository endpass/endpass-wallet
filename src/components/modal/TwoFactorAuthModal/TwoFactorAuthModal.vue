<template>
  <v-modal class="is-dark two-factor-auth-modal" @close="close">
    <template slot="header">
      {{ $t('components.twoFactorAuthModal.header') }}
    </template>
    <two-factor-auth-form
      :secret="secret"
      :email="email"
      :is-loading="isLoading"
      @submit="onFormSubmit"
    />
  </v-modal>
</template>

<script>
import TwoFactorAuthForm from './TwoFactorAuthForm';

export default {
  name: 'TwoFactorAuthModal',

  props: {
    secret: {
      type: String,
      default: '',
    },
    email: {
      type: String,
      default: '',
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },

  methods: {
    onFormSubmit({ code, verificationCode }) {
      this.$emit('confirm', {
        verificationCode,
        code,
      });
    },

    close() {
      this.$emit('close');
    },
  },

  components: {
    TwoFactorAuthForm,
  },
};
</script>

<style lang="scss">
.two-factor-auth-modal-section {
  margin-bottom: 30px;
}
</style>
