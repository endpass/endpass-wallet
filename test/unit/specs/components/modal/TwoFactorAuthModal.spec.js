import { shallow } from '@vue/test-utils';
import TwoFactorAuthModal from '@/components/modal/TwoFactorAuthModal';
import { generateStubs } from '@/utils/testUtils';

describe('TwoFactorAuthModal', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(TwoFactorAuthModal, {
      stubs: generateStubs(TwoFactorAuthModal),
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('two-factor-auth-modal');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render without secret and QR code', () => {
      expect(wrapper.element).toMatchSnapshot();

      wrapper.setProps({
        email: null,
        secret: 'secret',
      });

      expect(wrapper.element).toMatchSnapshot();

      wrapper.setProps({
        email: 'email',
        secret: null,
      });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render with secret and QR code', () => {
      wrapper.setProps({
        email: 'email',
        secret: 'secret',
      });

      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
