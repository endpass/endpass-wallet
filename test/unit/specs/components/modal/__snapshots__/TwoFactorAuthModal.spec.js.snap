// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwoFactorAuthModal render should render with secret and QR code 1`] = `
<v-modal>
   
  <p
    class="lead"
  >
    To protect your account, we strongly recommend ebaling
  two factor authentication. Once enabled, you will be able to log in by
  entering a code generated by an authenticator app.
  </p>
   
  <p
    class="lead"
  >
    Scan the QR code below into 
    <strong>
      Google
      Authenticator
    </strong>
    . If you are unable to scan the code, you can
  also enter the secret manually.
  </p>
   
  <v-form>
    <div
      class="field"
    >
      <div
        class="has-text-centered"
      >
        <img
          src="https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=otpauth://totp/Endpass:email?secret=secret"
        />
      </div>
       
      <p
        class="lead"
      >
        Write down the recovery secret below:
      </p>
       
      <label
        class="label"
      >
        Secret
      </label>
      
      secret
    
    </div>
     
    <p
      class="lead"
    >
      Enter the 6 digit code generated by your
      authenticator app.
    </p>
     
    <v-input
      label="Verification Code"
      name="verificationCode"
      size="6"
      type="number"
      validator="required|digits:6"
    />
     
    <v-button
      classname="is-primary is-medium"
    >
      Verify
    </v-button>
  </v-form>
</v-modal>
`;

exports[`TwoFactorAuthModal render should render without secret and QR code 1`] = `
<v-modal>
   
  <p
    class="lead"
  >
    To protect your account, we strongly recommend ebaling
  two factor authentication. Once enabled, you will be able to log in by
  entering a code generated by an authenticator app.
  </p>
   
  <p
    class="lead"
  >
    Scan the QR code below into 
    <strong>
      Google
      Authenticator
    </strong>
    . If you are unable to scan the code, you can
  also enter the secret manually.
  </p>
   
  <v-form>
    <!---->
     
    <p
      class="lead"
    >
      Enter the 6 digit code generated by your
      authenticator app.
    </p>
     
    <v-input
      label="Verification Code"
      name="verificationCode"
      size="6"
      type="number"
      validator="required|digits:6"
    />
     
    <v-button
      classname="is-primary is-medium"
    >
      Verify
    </v-button>
  </v-form>
</v-modal>
`;

exports[`TwoFactorAuthModal render should render without secret and QR code 2`] = `
<v-modal>
   
  <p
    class="lead"
  >
    To protect your account, we strongly recommend ebaling
  two factor authentication. Once enabled, you will be able to log in by
  entering a code generated by an authenticator app.
  </p>
   
  <p
    class="lead"
  >
    Scan the QR code below into 
    <strong>
      Google
      Authenticator
    </strong>
    . If you are unable to scan the code, you can
  also enter the secret manually.
  </p>
   
  <v-form>
    <!---->
     
    <p
      class="lead"
    >
      Enter the 6 digit code generated by your
      authenticator app.
    </p>
     
    <v-input
      label="Verification Code"
      name="verificationCode"
      size="6"
      type="number"
      validator="required|digits:6"
    />
     
    <v-button
      classname="is-primary is-medium"
    >
      Verify
    </v-button>
  </v-form>
</v-modal>
`;

exports[`TwoFactorAuthModal render should render without secret and QR code 3`] = `
<v-modal>
   
  <p
    class="lead"
  >
    To protect your account, we strongly recommend ebaling
  two factor authentication. Once enabled, you will be able to log in by
  entering a code generated by an authenticator app.
  </p>
   
  <p
    class="lead"
  >
    Scan the QR code below into 
    <strong>
      Google
      Authenticator
    </strong>
    . If you are unable to scan the code, you can
  also enter the secret manually.
  </p>
   
  <v-form>
    <!---->
     
    <p
      class="lead"
    >
      Enter the 6 digit code generated by your
      authenticator app.
    </p>
     
    <v-input
      label="Verification Code"
      name="verificationCode"
      size="6"
      type="number"
      validator="required|digits:6"
    />
     
    <v-button
      classname="is-primary is-medium"
    >
      Verify
    </v-button>
  </v-form>
</v-modal>
`;
